@claude =================================== FAILURES ===================================
_____________________________ TestRefer.test_refer _____________________________

self = <tests.test_main.TestRefer testMethod=test_refer>
m_tomato = <MagicMock name='get_book' id='139867417603984'>
m_calibre_title = <MagicMock name='get_book_by_title' id='139867417604320'>
m_calibre_isbn = <MagicMock name='get_book_by_isbn' id='139867417604656'>
m7 = <MagicMock name='_youshu' id='139867417604992'>
m6 = <MagicMock name='get_cover' id='139867417605328'>
m5 = <MagicMock name='_baike' id='139867417605664'>
m4 = <MagicMock name='get_cover' id='139867417606000'>
m3 = <MagicMock name='get_book_by_id' id='139867417606336'>
m2 = <MagicMock name='get_book_by_isbn' id='139867417606672'>
m1 = <MagicMock name='search_books' id='139867417607008'>

    @mock.patch("webserver.plugins.meta.douban.DoubanBookApi.search_books")
    @mock.patch("webserver.plugins.meta.douban.DoubanBookApi.get_book_by_isbn")
    @mock.patch("webserver.plugins.meta.douban.DoubanBookApi.get_book_by_id")
    @mock.patch("webserver.plugins.meta.douban.DoubanBookApi.get_cover")
    @mock.patch("webserver.plugins.meta.baike.BaiduBaikeApi._baike")
    @mock.patch("webserver.plugins.meta.baike.BaiduBaikeApi.get_cover")
    @mock.patch("webserver.plugins.meta.youshu.YoushuApi._youshu")
    @mock.patch("webserver.plugins.meta.calibre.CalibreMetadataApi.get_book_by_isbn")
    @mock.patch("webserver.plugins.meta.calibre.CalibreMetadataApi.get_book_by_title")
    @mock.patch("webserver.plugins.meta.tomato.TomatoNovelApi.get_book")
    def test_refer(self, m_tomato, m_calibre_title, m_calibre_isbn, m7, m6, m5, m4, m3, m2, m1):
        from tests.test_baike import BAIKE_PAGE
        from tests.test_douban import DOUBAN_BOOK, DOUBAN_SEARCH
        from tests.test_youshu import YOUSHU_PAGE
    
        m1.return_value = DOUBAN_SEARCH["books"]
        m2.return_value = dict(DOUBAN_BOOK)
        m3.return_value = dict(DOUBAN_BOOK)
        m4.return_value = ("jpg", b"image-body")
    
        m5.return_value = BAIKE_PAGE
        m6.return_value = ("jpg", b"image-body")
    
        m7.return_value = YOUSHU_PAGE
        m_calibre_isbn.return_value = []
        m_calibre_title.return_value = []
        m_tomato.return_value = None
    
        # with mock.patch("plugins.meta.baike.BaiduBaikeApi.get_book", return_value=self.fake_baidu) as m:
        # main.CONF["douban_baseurl"] = self.douban_url
        d = self.json("/api/book/1/refer")
        self.assertEqual(d["err"], "ok")
    
        global _app
        with mock.patch.object(_app.settings["legacy"], "set_metadata", return_value="Yo"):
            for book in d["books"]:
                body = "provider_key=%(provider_key)s&provider_value=%(provider_value)s" % book
>               r = self.json("/api/book/1/refer", method="POST", raise_error=True, body=body)

tests/test_main.py:767: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_main.py:178: in json
    rsp = self.fetch(url, *args, **kwargs)
/usr/local/lib/python3.13/dist-packages/tornado/testing.py:446: in fetch
    return self.io_loop.run_sync(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tornado.platform.asyncio.AsyncIOLoop object at 0x7f357085c4d0>
func = <function AsyncHTTPTestCase.fetch.<locals>.<lambda> at 0x7f3570881f80>
timeout = 60.0

    def run_sync(self, func: Callable, timeout: Optional[float] = None) -> Any:
        """Starts the `IOLoop`, runs the given function, and stops the loop.
    
        The function must return either an awaitable object or
        ``None``. If the function returns an awaitable object, the
        `IOLoop` will run until the awaitable is resolved (and
        `run_sync()` will return the awaitable's result). If it raises
        an exception, the `IOLoop` will stop and the exception will be
        re-raised to the caller.
    
        The keyword-only argument ``timeout`` may be used to set
        a maximum duration for the function.  If the timeout expires,
        a `asyncio.TimeoutError` is raised.
    
        This method is useful to allow asynchronous calls in a
        ``main()`` function::
    
            async def main():
                # do stuff...
    
            if __name__ == '__main__':
                IOLoop.current().run_sync(main)
    
        .. versionchanged:: 4.3
           Returning a non-``None``, non-awaitable value is now an error.
    
        .. versionchanged:: 5.0
           If a timeout occurs, the ``func`` coroutine will be cancelled.
    
        .. versionchanged:: 6.2
           ``tornado.util.TimeoutError`` is now an alias to ``asyncio.TimeoutError``.
        """
        if typing.TYPE_CHECKING:
            FutureCell = TypedDict(  # noqa: F841
                "FutureCell", {"future": Optional[Future], "timeout_called": bool}
            )
        future_cell = {"future": None, "timeout_called": False}  # type: FutureCell
    
        def run() -> None:
            try:
                result = func()
                if result is not None:
                    from tornado.gen import convert_yielded
    
                    result = convert_yielded(result)
            except Exception:
                fut = Future()  # type: Future[Any]
                future_cell["future"] = fut
                future_set_exc_info(fut, sys.exc_info())
            else:
                if is_future(result):
                    future_cell["future"] = result
                else:
                    fut = Future()
                    future_cell["future"] = fut
                    fut.set_result(result)
            assert future_cell["future"] is not None
            self.add_future(future_cell["future"], lambda future: self.stop())
    
        self.add_callback(run)
        if timeout is not None:
    
            def timeout_callback() -> None:
                # signal that timeout is triggered
                future_cell["timeout_called"] = True
                # If we can cancel the future, do so and wait on it. If not,
                # Just stop the loop and return with the task still pending.
                # (If we neither cancel nor wait for the task, a warning
                # will be logged).
                assert future_cell["future"] is not None
                if not future_cell["future"].cancel():
                    self.stop()
    
            timeout_handle = self.add_timeout(self.time() + timeout, timeout_callback)
        self.start()
        if timeout is not None:
            self.remove_timeout(timeout_handle)
        assert future_cell["future"] is not None
        if future_cell["future"].cancelled() or not future_cell["future"].done():
            if future_cell["timeout_called"]:
>               raise TimeoutError("Operation timed out after %s seconds" % timeout)
E               TimeoutError: Operation timed out after 60.0 seconds

/usr/local/lib/python3.13/dist-packages/tornado/ioloop.py:542: TimeoutError
------------------------------ Captured log call -------------------------------
ERROR    root:book.py:248 baidu 查询失败：too many values to unpack (expected 2)
ERROR    root:api.py:237 Get cover error: HTTPSConnectionPool(host='img8.xinhuashudian.com', port=443): Max retries exceeded with url: /images/2022/04/29/5276301c-0be0-47b8-9f82-2696efde3b7e.jpg (Caused by ConnectTimeoutError(<HTTPSConnection(host='img8.xinhuashudian.com', port=443) at 0x7f3571169590>, 'Connection to img8.xinhuashudian.com timed out. (connect timeout=10)'))
ERROR    tornado.application:concurrent.py:218 Exception after Future was cancelled
tornado.simple_httpclient.HTTPTimeoutError: Timeout during request

--- a/tests/test_main.py
+++ b/tests/test_main.py
@@ -737,7 +737,8 @@ class TestRefer(TestWithUserLogin):
     @mock.patch("webserver.plugins.meta.calibre.CalibreMetadataApi.get_book_by_isbn")
     @mock.patch("webserver.plugins.meta.calibre.CalibreMetadataApi.get_book_by_title")
     @mock.patch("webserver.plugins.meta.tomato.TomatoNovelApi.get_book")
-    def test_refer(self, m_tomato, m_calibre_title, m_calibre_isbn, m7, m6, m5, m4, m3, m2, m1):
+    @mock.patch("webserver.plugins.meta.xhsd.XhsdBookApi.get_book")
+    def test_refer(self, m_xhsd, m_tomato, m_calibre_title, m_calibre_isbn, m7, m6, m5, m4, m3, m2, m1):
         from tests.test_baike import BAIKE_PAGE
         from tests.test_douban import DOUBAN_BOOK, DOUBAN_SEARCH
         from tests.test_youshu import YOUSHU_PAGE
@@ -754,6 +755,7 @@ class TestRefer(TestWithUserLogin):
         m_calibre_isbn.return_value = []
         m_calibre_title.return_value = []
         m_tomato.return_value = None
+        m_xhsd.return_value = None
 
         # with mock.patch("plugins.meta.baike.BaiduBaikeApi.get_book", return_value=self.fake_baidu) as m:
         # main.CONF["douban_baseurl"] = self.douban_url

